stonesoup/hypothesiser/probability.py

```diff
@@ -126,18 +146,57 @@ def hypothesise(self, track, detections, timestamp, **kwargs):
                 prediction, detection.measurement_model, **kwargs)
             # Calculate difference before to handle custom types (mean defaults to zero)
             # This is required as log pdf coverts arrays to floats
-            log_pdf = mn.logpdf(
+            log_pdf = multivariate_normal.logpdf(
                 (detection.state_vector - measurement_prediction.state_vector).ravel(),
                 cov=measurement_prediction.covar)
             pdf = Probability(log_pdf, log_value=True)
-            probability = (pdf * self.prob_detect)/self.clutter_spatial_density
 
-            # True detection hypothesis
-            hypotheses.append(
-                SingleProbabilityHypothesis(
-                    prediction,
-                    detection,
-                    probability,
-                    measurement_prediction))
+            if self._is_valid_measurement(measurement_prediction, detection):
+                validated_measurements += 1
+                valid_measurement = True
+
+            if self.include_all or valid_measurement:
+                probability = pdf * self.prob_detect
+                if self.clutter_spatial_density is None:
+                    # Note: will divide by validated measurements count later...
+                    probability *= self._validation_region_volume(
+                        self.prob_gate, measurement_prediction)
+                else:
+                    probability /= self.clutter_spatial_density
+
+                # True detection hypothesis
+                hypotheses.append(
+                    SingleProbabilityHypothesis(
+                        prediction,
+                        detection,
+                        probability,
+                        measurement_prediction))
+
+        if self.clutter_spatial_density is None:
+            for hypothesis in hypotheses[1:]:  # Skip missed detection
+                hypothesis.probability /= validated_measurements
 
         return MultipleHypothesis(hypotheses, normalise=True, total_weight=1)
+
+    def _is_valid_measurement(self, meas_pred, det):
+        z = meas_pred.state_vector - det.state_vector
+        return \
+            z.T@self._inv_cov(meas_pred)@z <= self._gate_threshold(self.prob_gate, meas_pred.ndim)
+
+    @classmethod
+    @lru_cache()
+    def _validation_region_volume(cls, prob_gate, meas_pred):
+        n = meas_pred.ndim
+        gate_threshold = cls._gate_threshold(prob_gate, n)
+        c_z = np.pi**(n/2) / gamma(n/2 + 1)
+        return c_z * gate_threshold**(n/2) * np.sqrt(det(meas_pred.covar))
+
+    @staticmethod
+    @lru_cache()
+    def _inv_cov(meas_pred):
```

Comment:
Not too keen on single-line functions, but I'm guessing this is done to take advantage of caching?

Reply:
Yes. However, there are other modifications with Mahalanobis measure that include built in caching. I may rebase of main and then can take advantage of that and remove the need for this function.